File: app-geekhome/src/main/java/eu/geekhome/rest/instances/InstancesController.java
package eu.geekhome.rest.instances;

import eu.geekhome.services.configurable.Configurable;
import eu.geekhome.services.configurable.FieldDefinition;
import eu.geekhome.services.configurable.FieldValidationResult;
import eu.geekhome.rest.PluginsCoordinator;
import eu.geekhome.services.repository.InstanceDto;
import eu.geekhome.services.repository.Repository;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.MediaType;
import java.util.*;

@Path("instances")
public class InstancesController {

    private Repository _repository;
    private final PluginsCoordinator _pluginsCoordinator;

    private Optional<Configurable> findConfigurable(String clazz) {
        return _pluginsCoordinator
                .getConfigurables()
                .stream()
                .filter((x) -> x.getClass().getSimpleName().equals(clazz))
                .findFirst();
    }

    @Inject
    public InstancesController(PluginsCoordinator pluginsCoordinator) {
        _pluginsCoordinator = pluginsCoordinator;
        try {
            _repository = pluginsCoordinator.getRepositories().get(0);
        } catch (Exception ex) {
            _repository = pluginsCoordinator.getRepositories().get(0);
        }
    }

    @SuppressWarnings("rawtypes")
    @POST
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public Map<String, FieldValidationResult> postInstances(InstanceDto instanceDto) throws Exception {
        Optional<Configurable> configurable = findConfigurable(instanceDto.getClazz());
        if (configurable.isPresent()) {
            Map<String, FieldValidationResult> validationResult = new HashMap<>();
            boolean isObjectValid = true;

            for (FieldDefinition fieldDefinition : configurable.get().getFieldDefinitions()) {
                String fieldValue = instanceDto.getFields().get(fieldDefinition.getName());
                FieldValidationResult isValid = fieldDefinition.validate(fieldValue);
                validationResult.put(fieldDefinition.getName(), isValid);
                if (!isValid.isValid()) {
                    isObjectValid = false;
                }
            }

            if (isObjectValid) {
                _repository.saveInstance(instanceDto);
            }

            return validationResult;
        } else {
            throw new Exception("Unsupported configurable class");
        }
    }

    @SuppressWarnings("rawtypes")
    @PUT
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public Map<String, FieldValidationResult> putInstances(InstanceDto instanceDto) throws Exception {
        Optional<Configurable> configurable = findConfigurable(instanceDto.getClazz());
        if (configurable.isPresent()) {
            Map<String, FieldValidationResult> validationResult = new HashMap<>();
            boolean isObjectValid = true;

            for (FieldDefinition fieldDefinition : configurable.get().getFieldDefinitions()) {
                String fieldValue = instanceDto.getFields().get(fieldDefinition.getName());
                FieldValidationResult isValid = fieldDefinition.validate(fieldValue);
                validationResult.put(fieldDefinition.getName(), isValid);
                if (!isValid.isValid()) {
                    isObjectValid = false;
                }
            }

            if (isObjectValid) {
                _repository.updateInstance(instanceDto);
            }

            return validationResult;
        } else {
            throw new Exception("Unsupported configurable class");
        }
    }

    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public InstanceDto getInstancesById(@PathParam("id") long id) {
        return _repository.getInstance(id);
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public List<InstanceDto> getInstancesOfClass(@QueryParam("class") String clazz) {
        if (clazz != null) {
            return _repository.getInstancesOfClazz(clazz);
        }

        return new ArrayList<>();
    }

    @DELETE
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    public void deleteInstance(@PathParam("id") long id) {
        _repository.deleteInstance(id);
    }
}

File: app-geekhome/src/main/java/eu/geekhome/rest/instances/InstancesController.kt
package eu.geekhome.rest.instances

import eu.geekhome.data.Repository
import eu.geekhome.data.instances.InstanceDto
import eu.geekhome.domain.extensibility.PluginsCoordinator
import eu.geekhome.rest.settings.ValidationResultMap
import eu.geekhome.domain.configurable.ConfigurableWithFields
import eu.geekhome.domain.configurable.FieldValidationResult
import java.util.*
import javax.inject.Inject
import javax.ws.rs.*
import javax.ws.rs.core.MediaType

@Path("instances")
class InstancesController @Inject constructor(
    private val pluginsCoordinator: PluginsCoordinator,
    private val repository: Repository
) {

    private fun findConfigurable(clazz: String): ConfigurableWithFields? {
        return pluginsCoordinator
            .configurables
            .filter { x -> x.javaClass.name.equals(clazz) }
            .filterIsInstance<ConfigurableWithFields>()
            .firstOrNull()
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    @Throws(Exception::class)
    fun postInstances(instanceDto: InstanceDto): ValidationResultMap {
        return validate(instanceDto) {
            repository.saveInstance(instanceDto)
        }
    }

    @PUT
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    @Throws(Exception::class)
    fun putInstances(instanceDto: InstanceDto): ValidationResultMap {
        return validate(instanceDto) {
            repository.updateInstance(instanceDto)
        }
    }

    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    fun getInstancesById(@PathParam("id") id: Long): InstanceDto {
        return repository.getInstance(id)
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    fun getInstancesOfClass(@QueryParam("class") clazz: String?): List<InstanceDto> {
        return if (clazz != null) {
            repository.getInstancesOfClazz(clazz)
        } else ArrayList()
    }

    @DELETE
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON + ";charset=utf-8")
    fun deleteInstance(@PathParam("id") id: Long) {
        repository.deleteInstance(id)
    }

    private fun validate(instanceDto: InstanceDto, onValidCallback: () -> Unit):
            MutableMap<String, FieldValidationResult> {
        val configurable = findConfigurable(instanceDto.clazz)
        return if (configurable != null) {
            val validationResult: MutableMap<String, FieldValidationResult> = HashMap()
            var isObjectValid = true
            for (fieldDefinition in configurable.fieldDefinitions.values) {
                val fieldValue = instanceDto.fields[fieldDefinition.name]
                val isValid = fieldDefinition.validate(fieldValue)
                validationResult[fieldDefinition.name] = isValid
                if (!isValid.valid) {
                    isObjectValid = false
                }
            }
            if (isObjectValid) {
                onValidCallback()
            }
            validationResult
        } else {
            throw Exception("Unsupported configurable class")
        }
    }
}
